common/network-common/src/main/java/org/apache/spark/network/sasl/aes/AesConfigMessage.java

```diff
@@ -0,0 +1,124 @@
+
+package org.apache.spark.network.sasl.aes;
+
+import java.nio.ByteBuffer;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+
+import org.apache.spark.network.protocol.Encodable;
+import org.apache.spark.network.protocol.Encoders;
+
+/**
+ * The AES cipher options for SASL encryption negotiation.
+ */
+public class AesConfigMessage implements Encodable {
+  /** Serialization tag used to catch incorrect payloads. */
+  private static final byte TAG_BYTE = (byte) 0xEB;
+
+  public int keySize;
+  public byte[] inKey;
+  public byte[] outKey;
+  public byte[] inIv;
+  public byte[] outIv;
+
+  public AesConfigMessage(int keySize, byte[] inKey, byte[] inIv, byte[] outKey, byte[] outIv) {
+    this.keySize = keySize;
+    this.inKey = inKey;
+    this.inIv = inIv;
+    this.outKey = outKey;
+    this.outIv = outIv;
+  }
+
+  /**
+   * Set key and input vector for cipher option
+   * @param keySize the size of key in byte.
+   * @param inKey The decrypt key of one side.
+   * @param inIv The input vector of one side.
+   * @param outKey The decrypt key of another side.
+   * @param outIv The input vector of another side.
+   */
+  public void setParameters(int keySize, byte[] inKey, byte[] inIv, byte[] outKey, byte[] outIv) {
+    this.keySize = keySize;
+    this.inKey = inKey;
+    this.inIv = inIv;
+    this.outKey = outKey;
+    this.outIv = outIv;
+  }
+
+  @Override
+  public int encodedLength() {
+    return 1 + 4 + ((inKey != null && inIv != null && outKey != null && outIv != null) ?
+      Encoders.ByteArrays.encodedLength(inKey) + Encoders.ByteArrays.encodedLength(inKey) +
+      Encoders.ByteArrays.encodedLength(inIv) + Encoders.ByteArrays.encodedLength(outIv) : 0);
+  }
+
+  @Override
+  public void encode(ByteBuf buf) {
+    buf.writeByte(TAG_BYTE);
+    buf.writeInt(keySize);
+    if (inKey != null && inIv != null && outKey != null && outIv != null) {
+      Encoders.ByteArrays.encode(buf, inKey);
+      Encoders.ByteArrays.encode(buf, inIv);
+      Encoders.ByteArrays.encode(buf, outKey);
+      Encoders.ByteArrays.encode(buf, outIv);
+    }
+  }
+
+  /**
+   * Encode the config message.
+   * @return ByteBuffer which contains encoded config message.
+   */
+  public ByteBuffer encodeMessage(){
+    ByteBuffer buf = ByteBuffer.allocate(encodedLength());
+
+    ByteBuf wrappedBuf = Unpooled.wrappedBuffer(buf);
+    wrappedBuf.clear();
+    encode(wrappedBuf);
+
+    return buf;
+  }
+
+  /**
+   * Decode the config message from buffer
+   * @param buffer the buffer contain encoded config message
+   * @return config message
+   */
+  public static AesConfigMessage decodeMessage(ByteBuffer buffer) {
+    ByteBuf buf = Unpooled.wrappedBuffer(buffer);
+
+    if (buf.readByte() != TAG_BYTE) {
+      throw new IllegalStateException("Expected SaslMessage, received something else"
+        + " (maybe your client does not have SASL enabled?)");
+    }
+
+    int keySize = buf.readInt();
+
+    if (buf.readableBytes() > 0) {
```

Comment:
So... you need this check because the client is sending an empty `AesConfigMessage` to the server, which then creates the keys and sends a response back to the client. Instead, why not save the round trip and have the client generate the keys and send a single message to the server, without requiring a response? Then the overloading of `AesConfigMessage` as both request and response goes away, and things become less confusing.

(Hint: see the `TransportClient.send()` method.)